File: test/language/constraints/cpp/ChoiceConstraintsTest.cpp
#include "constraints/choice_constraints/ChoiceConstraints.h"
#include "gtest/gtest.h"
#include "zserio/BitStreamReader.h"
#include "zserio/BitStreamWriter.h"
#include "zserio/CppRuntimeException.h"

namespace constraints
{
namespace choice_constraints
{

class ChoiceConstraintsTest : public ::testing::Test
{
protected:
    void writeChoiceConstraintsToByteArray(
            zserio::BitStreamWriter& writer, bool selector, uint8_t value8, uint16_t value16)
    {
        if (selector)
            writer.writeBits(value8, 8);
        else
            writer.writeBits(value16, 16);
    }

    static const uint8_t VALUE8_CORRECT_CONSTRAINT;
    static const uint8_t VALUE8_WRONG_CONSTRAINT;

    static const uint16_t VALUE16_CORRECT_CONSTRAINT;
    static const uint16_t VALUE16_WRONG_CONSTRAINT;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const uint8_t ChoiceConstraintsTest::VALUE8_CORRECT_CONSTRAINT = 1;
const uint8_t ChoiceConstraintsTest::VALUE8_WRONG_CONSTRAINT = 0;

const uint16_t ChoiceConstraintsTest::VALUE16_CORRECT_CONSTRAINT = 256;
const uint16_t ChoiceConstraintsTest::VALUE16_WRONG_CONSTRAINT = 255;

TEST_F(ChoiceConstraintsTest, readConstructorCorrectConstraints)
{
    const bool selector = true;
    const uint8_t value8 = VALUE8_CORRECT_CONSTRAINT;
    zserio::BitStreamWriter writer(bitBuffer);
    writeChoiceConstraintsToByteArray(writer, selector, value8, 0);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    ChoiceConstraints choiceConstraints(reader, selector);

    ASSERT_EQ(selector, choiceConstraints.getSelector());
    ASSERT_EQ(value8, choiceConstraints.getValue8());
}

TEST_F(ChoiceConstraintsTest, readConstructorWrongValue8Constraint)
{
    const bool selector = true;
    const uint8_t value8 = VALUE8_WRONG_CONSTRAINT;
    zserio::BitStreamWriter writer(bitBuffer);
    writeChoiceConstraintsToByteArray(writer, selector, value8, 0);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    ASSERT_THROW(ChoiceConstraints choiceConstraints(reader, selector), zserio::CppRuntimeException);
}

TEST_F(ChoiceConstraintsTest, readConstructorWrongValue16Constraint)
{
    const bool selector = false;
    const uint16_t value16 = VALUE16_WRONG_CONSTRAINT;
    zserio::BitStreamWriter writer(bitBuffer);
    writeChoiceConstraintsToByteArray(writer, selector, 0, value16);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    ASSERT_THROW(ChoiceConstraints choiceConstraints(reader, selector), zserio::CppRuntimeException);
}

TEST_F(ChoiceConstraintsTest, writeCorrectConstraints)
{
    const bool selector = false;
    const uint16_t value16 = VALUE16_CORRECT_CONSTRAINT;
    ChoiceConstraints choiceConstraints;
    choiceConstraints.initialize(selector);
    choiceConstraints.setValue16(value16);

    zserio::BitStreamWriter writer(bitBuffer);
    choiceConstraints.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    const ChoiceConstraints readChoiceConstraints(reader, selector);
    ASSERT_EQ(selector, readChoiceConstraints.getSelector());
    ASSERT_EQ(value16, readChoiceConstraints.getValue16());
    ASSERT_TRUE(choiceConstraints == readChoiceConstraints);
}

TEST_F(ChoiceConstraintsTest, writeWrongValue8Constraint)
{
    const bool selector = true;
    const uint8_t value8 = VALUE8_WRONG_CONSTRAINT;
    ChoiceConstraints choiceConstraints;
    choiceConstraints.initialize(selector);
    choiceConstraints.setValue8(value8);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(choiceConstraints.write(writer), zserio::CppRuntimeException);
}

TEST_F(ChoiceConstraintsTest, writeWrongValue16Constraint)
{
    const bool selector = false;
    const uint16_t value16 = VALUE16_WRONG_CONSTRAINT;
    ChoiceConstraints choiceConstraints;
    choiceConstraints.initialize(selector);
    choiceConstraints.setValue16(value16);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(choiceConstraints.write(writer), zserio::CppRuntimeException);
}

} // namespace choice_constraints
} // namespace constraints

File: test/language/constraints/python/ChoiceConstraintsTest.py
import unittest
import zserio

from testutils import getZserioApi

class ChoiceConstraintsTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "constraints.zs").choice_constraints

    def testReadCorrectConstraints(self):
        selector = True
        value8 = self.VALUE8_CORRECT_CONSTRAINT
        writer = zserio.BitStreamWriter()
        self.__class__._write(writer, selector, value8, 0)
        reader = zserio.BitStreamReader(writer.getByteArray())

        choiceConstraints = self.api.ChoiceConstraints(selector)
        choiceConstraints.read(reader)
        self.assertEqual(selector, choiceConstraints.getSelector())
        self.assertEqual(value8, choiceConstraints.getValue8())

    def testReadWrongValue8Constraint(self):
        selector = True
        value8 = self.VALUE8_WRONG_CONSTRAINT
        writer = zserio.BitStreamWriter()
        self.__class__._write(writer, selector, value8, 0)
        reader = zserio.BitStreamReader(writer.getByteArray())

        choiceConstraints = self.api.ChoiceConstraints(selector)
        with self.assertRaises(zserio.PythonRuntimeException):
            choiceConstraints.read(reader)

    def testReadWrongValue16Constraint(self):
        selector = False
        value16 = self.VALUE16_WRONG_CONSTRAINT
        writer = zserio.BitStreamWriter()
        self.__class__._write(writer, selector, 0, value16)
        reader = zserio.BitStreamReader(writer.getByteArray())

        choiceConstraints = self.api.ChoiceConstraints(selector)
        with self.assertRaises(zserio.PythonRuntimeException):
            choiceConstraints.read(reader)

    def testWriteCorrectConstraints(self):
        selector = False
        value16 = self.VALUE16_CORRECT_CONSTRAINT
        choiceConstraints = self.api.ChoiceConstraints(selector)
        choiceConstraints.setValue16(value16)

        writer = zserio.BitStreamWriter()
        choiceConstraints.write(writer)
        reader = zserio.BitStreamReader(writer.getByteArray())
        readChoiceConstraints = self.api.ChoiceConstraints.fromReader(reader, selector)
        self.assertEqual(selector, readChoiceConstraints.getSelector())
        self.assertEqual(value16, readChoiceConstraints.getValue16())
        self.assertEqual(choiceConstraints, readChoiceConstraints)

    def testWriteWrongValue8Constraint(self):
        selector = True
        value8 = self.VALUE8_WRONG_CONSTRAINT
        choiceConstraints = self.api.ChoiceConstraints(selector)
        choiceConstraints.setValue8(value8)

        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            choiceConstraints.write(writer)

    def testWriteWrongValue16Constraint(self):
        selector = False
        value16 = self.VALUE16_WRONG_CONSTRAINT
        choiceConstraints = self.api.ChoiceConstraints(selector)
        choiceConstraints.setValue16(value16)

        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            choiceConstraints.write(writer)

    @staticmethod
    def _write(writer, selector, value8, value16):
        writer.writeBits(value8 if selector else value16, 8)

    VALUE8_CORRECT_CONSTRAINT = 1
    VALUE8_WRONG_CONSTRAINT = 0

    VALUE16_CORRECT_CONSTRAINT = 256
    VALUE16_WRONG_CONSTRAINT = 255
